tests/bluetooth/bsim_bt/bsim_test_l2cap/src/main_l2cap_ecred.c

```diff
@@ -0,0 +1,563 @@
+		bs_trace_info_time(1, __VA_ARGS__); \
+	} while (0)
+
+static struct bt_conn *default_conn;
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+};
+
+#define DATA_MTU 2000
+#define DATA_MPS 65
+#define DATA_BUF_SIZE BT_L2CAP_SDU_BUF_SIZE(DATA_MTU)
+#define L2CAP_CHANNELS 2
+#define SERVERS 1
+#define SDU_SEND_COUNT 200
+
+NET_BUF_POOL_FIXED_DEFINE(rx_data_pool, L2CAP_CHANNELS, BT_L2CAP_BUF_SIZE(DATA_BUF_SIZE), 8, NULL);
+NET_BUF_POOL_FIXED_DEFINE(tx_data_pool, L2CAP_CHANNELS + 1, BT_L2CAP_BUF_SIZE(DATA_MTU), 8, NULL);
+
+static struct bt_l2cap_server servers[SERVERS];
+void send_sdu_chan_worker(struct k_work *item);
+
+static struct channel {
+	uint8_t chan_id; /* Internal number that identifies L2CAP channel. */
+	struct bt_l2cap_le_chan le;
+	bool in_use;
+	int sdus_received;
+	int bytes_to_send;
+	int itteration;
+	struct net_buf *buf;
+	struct k_work work;
+	uint8_t payload[DATA_MTU];
+} channels[L2CAP_CHANNELS];
+CREATE_FLAG(is_connected);
+#define MY_STACK_SIZE 512
+#define MY_PRIORITY 5
+
+K_THREAD_STACK_DEFINE(my_stack_area0, MY_STACK_SIZE);
+struct k_work_q my_work_q0;
+K_THREAD_STACK_DEFINE(my_stack_area1, MY_STACK_SIZE);
+struct k_work_q my_work_q1;
+
+static void init_workqs(void)
+{
+	k_work_queue_init(&my_work_q0);
+	k_work_queue_start(&my_work_q0, my_stack_area0,
+			   K_THREAD_STACK_SIZEOF(my_stack_area0), MY_PRIORITY,
+			   NULL);
+
+	k_work_queue_init(&my_work_q1);
+	k_work_queue_start(&my_work_q1, my_stack_area1,
+			   K_THREAD_STACK_SIZEOF(my_stack_area1), MY_PRIORITY,
+			   NULL);
+}
+
+static struct channel *get_free_channel(void)
+{
+	struct channel *chan;
+
+	for (int idx = 0; idx < L2CAP_CHANNELS; idx++) {
+		if (channels[idx].in_use) {
+			continue;
+		}
+		chan = &channels[idx];
+		(void)memset(chan, 0, sizeof(*chan));
+		chan->chan_id = idx;
+		channels[idx].in_use = true;
+		(void)memset(chan->payload, idx, sizeof(chan->payload));
+		k_work_init(&chan->work, send_sdu_chan_worker);
+		return chan;
+	}
+
+	return NULL;
+}
+
+static struct net_buf *chan_alloc_buf_cb(struct bt_l2cap_chan *chan)
+{
+	LOG_DBG("Allocated on chan %p", chan);
+	return net_buf_alloc(&rx_data_pool, K_FOREVER);
+}
+
+static int chan_recv_cb(struct bt_l2cap_chan *l2cap_chan, struct net_buf *buf)
+{
+	struct channel *chan = CONTAINER_OF(l2cap_chan, struct channel, le);
+	int received_iterration = *((int *)(buf->data));
+
+	LOG_DBG("received_iterration %i sdus_receied %i, chan_id: %d, data_length: %d", received_iterration,  chan->sdus_received, chan->chan_id, buf->len);
+	if (received_iterration != chan->sdus_received) {
+		FAIL("Received out of sequence data.");
+	}
+
+	int retval = memcmp(buf->data + sizeof(int), chan->payload + sizeof(int), buf->len - sizeof(int));
+
+	if (retval) {
+		FAIL("Payload received didn't match expected value memcmp returned %i", retval);
+	}
+
+	if (chan->chan_id == 0) {
+		/*By the time we rx on chan 0, we should have alrady received on chan1*/
```

Comment:
Not sure what this means. If we RX on chan 0, we have already RXed on chan 1? Do we send on chan 1, before chan 0?

Reply:
Ah, so the assumption here is then that if we send a smaller payload, then we will receive that that faster than the larger payload? That makes sense, but will this always be the case when taking scheduling into account, and possible other delays from processing data in the stack? 

If the test assumes and expects this, then I guess we should add a check to verify it every time, and call `FAIL` if something goes wrong.

Reply:
Isn't it already done here ?https://github.com/zephyrproject-rtos/zephyr/pull/41741/commits/7b488b346f5214e0c5663435d69681ee4236cc6d#diff-ccecebdd1f83e23529fe3da637ceac7187e5f83763841316577d2b37bb6dc9d2R101